### PR TITLE
RAC 5.0

### PR DESCRIPTION
**Pull request by mdiep**
* #2852 Make DynamicProperty generic and automatically support bridgeable value types
* #2863 [RAC 5.0] Tweaks for generic DynamicProperty
* #2888 Allow DynamicProperty to be initialized with an optional object
* #2922 Property composition operators
* #2963 [Bugfix] Property composition operator
* #2966 [5.0] Changes to producer- and signal-sourced `AnyProperty`s.
* #2984 Adopt `inout` parameter in `Atomic.modify`
* #3028 Isolate unavailable APIs in 5.0.
* #3001 [5.0] Swift 3.0 Groundwork.

### CODE REVIEW

**Comment by NachoSoto** [comment]
I'm going to do one last pass! Sorry in advance for the noise.

File: ReactiveCocoa/Swift/Deprecations+Removals.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// MARK: Removed APIs in ReactiveCocoa 5.0.
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Deprecations+Removals.swift]: Beautiful

File: ReactiveCocoa/Swift/Property.swift [hunk 2]
```diff
@@ -8,66 +8,437 @@ public protocol PropertyType {
...
+	/// `property` and `anotherProperty`. The resulting property captures `property`
+	/// and `anotherProperty`.
+	private init<P1: PropertyType, P2: PropertyType>(_ firstProperty: P1, _ secondProperty: P2, @noescape transform: SignalProducer<P1.Value, NoError> -> SignalProducer<P2.Value, NoError> -> SignalProducer<Value, NoError>) {
+		self.init(propertyProducer: transform(firstProperty.producer)(secondProperty.producer),
+		          capturing: AnyProperty.capture(firstProperty) + AnyProperty.capture(secondProperty))
+	}
+
+	/// Initializes a property from a producer that promises to send at least one
+	/// value synchronously in its start handler before sending any subsequent event.
+	/// If the producer fails its promise, a fatal error would be raised.
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Property.swift]: > from a producer that promises to send at least one value synchronously in its start handler before sending any subsequent event.

Huh? How is that "promise" guaranteed here?

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 3]
```diff
@@ -1507,3 +1350,124 @@ private final class DeallocationToken {
 		observer.sendCompleted()
 	}
 }
+
+extension SignalProducer {
+	private static func bufferingProducer(capacity: Int) -> (SignalProducer, Signal<Value, Error>.Observer) {
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Oh wow, I didn't realize we'd still need this.

File: ReactiveCocoaTests/Swift/PropertySpec.swift [hunk 4]
```diff
@@ -330,46 +334,803 @@ class PropertySpec: QuickSpec {
...
+						expect(latestValue) == 4
+					}
+
+					it("should retain its source property") {
+						var property = Optional(MutableProperty(1))
+						weak var weakProperty = property
+
+						var firstMappedProperty = Optional(property!.map { $0 + 1 })
+						var secondMappedProperty = Optional(firstMappedProperty!.map { $0 + 2 })
+						_ = secondMappedProperty
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/PropertySpec.swift]: What's this for?

File: ReactiveCocoaTests/Swift/PropertySpec.swift [hunk 5]
```diff
@@ -330,46 +334,803 @@ class PropertySpec: QuickSpec {
...
+
+						property = nil
+						expect(completed) == true
+					}
+				}
+			}
+
+			describe("flattening") {
+				describe("flatten") {
+					describe("FlattenStrategy.Concat") {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/PropertySpec.swift]: These 3 levels seem unnecessary, I think you can reduce it to just `flattening` and `FlattenStrategy.Concat`.

**Comment by NachoSoto** [comment]
Cool! A bunch of small things, but looking good :)

File: ReactiveCocoa/Swift/Property.swift (hunk 2 above)
Comment (andersio) [inline ReactiveCocoa/Swift/Property.swift]: No compile time guarantee. That's why it is private.

Comment (andersio) [inline ReactiveCocoa/Swift/Property.swift]: 2 alternatives though

1. Makes it behave like `first()` (use semaphore)
2. Introduces `withValue` to `PropertyType` that can implicitly expose `MutableProperty`'s lock. Uses it together with `signal`. While it worked, it was turned down by @mdiep though for needing nested lock.

Comment (andersio) [inline ReactiveCocoa/Swift/Property.swift]: Hmm I have an idea to make it statically safe. But it requires a change to `PropertyType`.

#3040

File: ReactiveCocoaTests/Swift/PropertySpec.swift (hunk 4 above)
Comment (andersio) [inline ReactiveCocoaTests/Swift/PropertySpec.swift]: Suppress the "written to but never read" warning.

Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/PropertySpec.swift]: Gotcha. Can you add a comment to clarify?

Comment (andersio) [inline ReactiveCocoaTests/Swift/PropertySpec.swift]: I will include it in #3023.

**Comment by mdiep** [comment]
This branch could use a merge from `master` if anyone wants to open a PR to do that.

**Comment by andersio** [comment]
(conflicts conflicts conflicts)

**Comment by mdiep** [comment]
Let's do this

**Pull request merged**